zio-http/src/main/scala/zio/http/netty/server/ServerInboundHandler.scala

```diff
@@ -28,216 +35,249 @@ private[zio] final case class ServerInboundHandler(
-      exit: HExit[Any, Throwable, Response],
-      jRequest: HttpRequest,
-      time: service.ServerTime,
-      runtime: NettyRuntime,
-    ): ZIO[Any, Throwable, Unit] = {
+        if (!attemptImmediateWrite(ctx, exit, time))
+          writeResponse(ctx, env, exit, jReq)(releaseRequest)
+        else
+          releaseRequest()
 
-      for {
-        response <- exit.toZIO.unrefine { case error => Option(error) }.catchAll {
-          case None        => ZIO.succeed(HttpError.NotFound(jRequest.uri()).toResponse)
-          case Some(error) => ZIO.succeed(HttpError.InternalServerError(cause = Some(error)).toResponse)
-        }
-        _        <-
-          if (response.isWebSocket) ZIO.attempt(upgradeToWebSocket(jRequest, response, runtime))
-          else
-            for {
-              jResponse <- response.encode()
-              _         <- ZIO.attemptUnsafe(implicit u => setServerTime(time, response, jResponse))
-              _         <- ZIO.attempt(ctx.writeAndFlush(jResponse))
-              flushed <- if (!jResponse.isInstanceOf[FullHttpResponse]) response.body.write(ctx) else ZIO.succeed(true)
-              _       <- ZIO.attempt(ctx.flush()).when(!flushed)
-            } yield ()
-
-        _ <- ZIO.attemptUnsafe(implicit u => setContentReadAttr(false))
-      } yield log.debug("Full write performed")
-    }
+      case jReq: HttpRequest =>
+        log.debug(s"HttpRequest: [${jReq.method()} ${jReq.uri()}]")
+        val req  = makeZioRequest(ctx, jReq)
+        val exit = http.execute(req)
 
-    def canHaveBody(jReq: HttpRequest): Boolean = {
-      jReq.method() == HttpMethod.TRACE ||
-      jReq.headers().contains(HttpHeaderNames.CONTENT_LENGTH) ||
-      jReq.headers().contains(HttpHeaderNames.TRANSFER_ENCODING)
-    }
+        if (!attemptImmediateWrite(ctx, exit, time)) {
 
-    def contentIsRead: Boolean =
-      ctx.channel().attr(isReadKey).get()
+          if (
+            jReq.method() == HttpMethod.TRACE ||
+            jReq.headers().contains(HttpHeaderNames.CONTENT_LENGTH) ||
+            jReq.headers().contains(HttpHeaderNames.TRANSFER_ENCODING)
+          )
+            ctx.channel().config().setAutoRead(false)
 
-    def hasChanged(r1: Response, r2: Response): Boolean =
-      (r1.status eq r2.status) && (r1.body eq r2.body) && (r1.headers eq r2.headers)
+          writeResponse(ctx, env, exit, jReq)(() => ctx.channel().config().setAutoRead(true): Unit)
 
-    def makeZioRequest(nettyReq: HttpRequest): Request = {
-      val nettyHttpVersion = nettyReq.protocolVersion()
-      val protocolVersion  = nettyHttpVersion match {
-        case HttpVersion.HTTP_1_0 => Version.Http_1_0
-        case HttpVersion.HTTP_1_1 => Version.Http_1_1
-        case _ => throw new IllegalArgumentException(s"Unsupported HTTP version: ${nettyHttpVersion}")
-      }
+        }
 
-      // TODO: We need to bring this back, probably not part of Request.
-      // val remoteAddress = ctx.channel().remoteAddress() match {
```

Comment:
`Http` can gain a new subtype that allows accessing channel-related information that is backend independent, such as remote address.